### PR TITLE
[MRV2][Spec Decode] DFlash

### PR DESCRIPTION
**Pull request by benchislett**
## Purpose

Implement DFlash for ModelRunnerV2, with full cudagraph support.

Performance result:
MRV1:
image
MRV2:
image

Running Qwen3 8B FP8 + DFlash on 1xGB200. ~1.2x speedup, from 5.2ms to 4.3ms per step (on independent benchmark, separate from profiling run).

## Testing

Ran existing DFlash correctness testing with MRV1 and MRV2, got pass and identical acceptance rates.

Enabled E2E DFlash correctness and AR regression tests for both MRV1 and MRV2, both passing locally

### CODE REVIEW

File: vllm/v1/worker/gpu/spec_decode/dflash/cudagraph.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    def __init__(
+        self,
+        vllm_config: VllmConfig,
+        device: torch.device,
+        cudagraph_mode: CUDAGraphMode,
+        decode_query_len: int,
+    ):
+        super().__init__(vllm_config, device, cudagraph_mode, decode_query_len)
+
+        # Use a dedicated pool for DFlash to avoid memory overlap with the main
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/cudagraph.py]: TODO: This may be out of date with main, and could be removed I think

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 2]
```diff
@@ -0,0 +1,616 @@
+# SPDX-License-Identifier: Apache-2.0
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py:1]: This file likely repeats a lot of code from `autoregressive/speculator.py`. Please review thoroughly

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 3]
```diff
@@ -0,0 +1,616 @@
...
+                : num_reqs_padded + 1
+            ],
+            query_start_loc_cpu=query_start_loc_cpu,
+            max_query_len=self.num_query_per_req,
+            seq_lens=self.input_buffers.seq_lens[:num_reqs_padded],
+            max_seq_len=self.draft_max_seq_len,
+            block_tables=block_tables,
+            slot_mappings=slot_mappings,
+            kv_cache_config=self.kv_cache_config,
+            causal=False,
```
Comment (benchislett) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: TODO: support causal dflash as introduced in https://github.com/vllm-project/vllm/pull/43445

File: vllm/config/vllm.py [hunk 4]
```diff
@@ -2008,12 +2008,21 @@ def _get_v2_model_runner_unsupported_features(self) -> list[str]:
             # TODO: ngram / ngram_gpu are not supported by the v2 model runner yet
             if speculative_config.method in ("ngram", "ngram_gpu"):
                 unsupported.append("ngram/ngram_gpu speculative decoding")
-            elif speculative_config.method not in ("eagle", "eagle3", "mtp"):
+            elif speculative_config.method not in (
+                "eagle",
+                "eagle3",
+                "mtp",
+                "dflash",
+            ):
```
Comment (WoosukKwon) [inline vllm/config/vllm.py]: nit
```suggestion
            elif speculative_config.method not in ("eagle", "eagle3", "mtp", "dflash"):
```

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py (hunk 3 above)
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: You can reuse the `_build_draft_attn_metadata` from `DraftModelSpeculator` here with some minor edits (adding `num_query_per_req` and `causal` parameters.

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 5]
```diff
@@ -0,0 +1,616 @@
...
+                temperature,
+                seeds,
+                positions + 1,
+                apply_temperature=True,
+                output_processed_logits=draft_logits,
+                output_processed_logits_col=draft_step,
+                use_fp64=self.use_fp64_gumbel,
+            )
+        else:
+            return logits.argmax(dim=-1)
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: nit: This method is pretty much the same as the one used by autoregressive speculator. We could add it to the `DraftModelSpeculator` class as a helper method for both to use.

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+    parallel_drafting_token_id: int,
+    num_query_per_req: int,
+    num_speculative_steps: int,
+    max_num_reqs: int,
+    max_num_tokens: int,
+) -> None:
+    num_reqs = input_batch.num_reqs
+    assert num_reqs > 0
+    # Cover the longest possible per-request span (ctx + query).
+    max_tokens_per_req = input_batch.num_tokens + num_query_per_req
```
Comment (TheEpicDolphin) [inline vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: Was it intentional to use `input_batch.num_tokens` here? I think that is the total number scheduled tokens for the decode step, which would make `max_tokens_per_req` larger than it needs to be.

**Review by TheEpicDolphin — COMMENTED** [review]
This code looks good overall. I left a few suggestions to reduce code duplications, and flagged a potential perf improvement

File: vllm/v1/worker/gpu/spec_decode/dflash/speculator.py (hunk 6 above)
Comment (benchislett) [reply vllm/v1/worker/gpu/spec_decode/dflash/speculator.py]: Oh, I see what you mean. This should instead be the maximum query length of the target batch, not the total number of tokens in the target batch. That's my mistake. Should be safe nevertheless as OOB indices just mask out and don't write any data.

I'll try to fix this.

File: tests/v1/e2e/spec_decode/test_spec_decode.py [hunk 7]
```diff
@@ -1294,19 +1294,24 @@ def dflash_config():
         },
         max_model_len=32768,
         max_num_seqs=128,
-        gpu_memory_utilization=0.85,
+        gpu_memory_utilization=0.75,
```
Comment (benchislett) [inline tests/v1/e2e/spec_decode/test_spec_decode.py]:
```suggestion
        gpu_memory_utilization=0.85,
```

**Review by TheEpicDolphin — APPROVED** [review]
Looks good to me, thank you! cc: @WoosukKwon

**Review by WoosukKwon — APPROVED** [review]
